build-aux/snap/snapcraft.yaml

```diff
@@ -28,172 +29,272 @@ license: GPL-3.0
-  # libc6 is part of core but we need it in the snapd snap for
-  # CommandFromSystemSnap
-  libc6:
-    plugin: nil
-    stage-packages:
-      - libc6
-      - libc-bin
-      - libgcc1
-    stage:
-      - lib/*
-      - usr/lib/*
-      - lib64/*
-      - etc/ld.so.conf
-      - etc/ld.so.conf.d/*
-    override-stage: |
-      snapcraftctl stage
-      # fix symlinks of ld.so to be relative
-      if [ "$(readlink -f lib64/ld-linux-x86-64.so.2)" = "/lib/x86_64-linux-gnu/ld-2.23.so" ]; then
-          ln -f -s ../lib/x86_64-linux-gnu/ld-2.23.so lib64/ld-linux-x86-64.so.2
-      fi
-      if [ "$(readlink -f lib64/ld64.so.2)" = "/lib/powerpc64le-linux-gnu/ld-2.23.so" ]; then
-          ln -f -s ../lib/powerpc64le-linux-gnu/ld-2.23.so lib64/ld64.so.2
-      fi
+      - -usr/lib/$CRAFT_ARCH_TRIPLET_BUILD_FOR/ld*.so*
+      - -lib32
+      - -lib64
+      - -usr/lib32
+      - -usr/lib64
+      - -etc/ld.so.conf.d
+      - -usr/share/man
+    override-build: |
+      craftctl default
+      mv "${CRAFT_PART_INSTALL}/usr/lib/libnss_extrausers.so.2" "${CRAFT_PART_INSTALL}/usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/"
+      cp -rT "${CRAFT_PART_INSTALL}/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}" "${CRAFT_PART_INSTALL}/usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}"
+      rm -rf "${CRAFT_PART_INSTALL}/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}"
+
   apparmor:
     plugin: autotools
-    build-packages: [bison, flex, gettext, g++, pkg-config, wget]
+    build-packages:
+      - bison
+      - flex
+      - gettext
+      - g++
+      - pkg-config
+      - wget
     source: https://launchpad.net/apparmor/3.0/3.0.8/+download/apparmor-3.0.8.tar.gz
     override-pull: |
-      snapcraftctl pull
+      craftctl default
       # add support for mediating posix mqueue's and userns - these come from
       # the ubuntu source package for lunar
       # https://git.launchpad.net/ubuntu/+source/apparmor/tree/debian/patches/ubuntu?h=ubuntu/lunar
       for feature in mqueue userns; do
         wget https://git.launchpad.net/ubuntu/+source/apparmor/plain/debian/patches/ubuntu/add-${feature}-support.patch?h=ubuntu/lunar -O - | patch -p1
       done
       # apply local apparmor patches
-      for patch in $SNAPCRAFT_PROJECT_DIR/build-aux/snap/local/patches/apparmor/*; do
+      for patch in "${CRAFT_PROJECT_DIR}"/build-aux/snap/local/patches/apparmor/*; do
           patch -p1 < $patch
       done
     override-build: |
-      cd $SNAPCRAFT_PART_BUILD/libraries/libapparmor
+      cd "${CRAFT_PART_BUILD}"/libraries/libapparmor
       ./autogen.sh
-      ./configure --prefix=/usr --disable-man-pages --disable-perl --disable-python --disable-ruby
-      make -j$(nproc)
-      # place libapparmor into staging area for use by snap-confine
-      make -C src install DESTDIR=$SNAPCRAFT_STAGE
-      cd $SNAPCRAFT_PART_BUILD/parser
+      ./configure --prefix=/usr --libdir="/usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}" --disable-man-pages --disable-perl --disable-python --disable-ruby
```

Comment:
This is unrelated to the move to 22.04 but I want to ask the same question about patches. Should we use the source package from Ubuntu instead of the upstream tarball so that we can apply ongoing security fixes this way?

Reply:
I agree. But I also wonder why we had to use this build of apparmor in the first place. So maybe we should ask who introduced this build.

Reply:
glibc is best tracking ubuntu source package; as upstream may not have security fixes yet.
apparmor is tracked from upstream, as it is intentionally way ahead of what even devel Ubuntu release ships as maintained by Ubuntu Security / Apparmor team. The intention being to allow snapd integrate new apparmor features even ahead of them being available on any kernels.

Reply:
Kind of related to the package sources. We currently use ubuntu/lunar instead of ubuntu/lunar-updates, I wonder what's the reason for this, unless lunar simply wasn't released yet at the time apparmor got added to snapd, hence there could have been no lunar-updates branch at that time yet.

Reply:
I suppose it is because it is specific patches. They are going to be modified only if they have to be "rebased". I do not think we would get "new versions" of the patches, with fixes in them. It is possible that if when we update the version of upstream apparmor, we need to pick some other rebased version of patches, maybe in other branches. But where, it does not matter.

Reply:
yes, this uses lunar so that we get the patches which match the upstream 3.0.8 tarball - things will change anyway when we move to apparmor 4.0 via  #13354